// react-imvc/controller/attachDevToolsIfPossible.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = attachDevToolsIfPossiable;
function attachDevToolsIfPossiable(store) {
  if (process.env.NODE_ENV === "production") {
    return;
  }
  if (typeof window === "undefined" || !window.__REDUX_DEVTOOLS_EXTENSION__) {
    return;
  }

  var __FROM_REDUX_DEVTOOLS_EXTENSION__ = "__FROM_REDUX_DEVTOOLS_EXTENSION__";

  var options = {
    name: window.location.pathname + window.location.search,
    actionsWhitelist: Object.keys(store.actions)
  };
  var reduxStore = __REDUX_DEVTOOLS_EXTENSION__(store.getState, store.getState(), options);
  var isSync = false;
  store.subscribe(function (data) {
    if (!data || data.actionType === __FROM_REDUX_DEVTOOLS_EXTENSION__) {
      return;
    }
    isSync = true;
    reduxStore.dispatch({
      type: data.actionType,
      payload: data.actionPayload
    });
    isSync = false;
  });

  reduxStore.subscribe(function () {
    if (!isSync) {
      store.replaceState(reduxStore.getState(), {
        actionType: __FROM_REDUX_DEVTOOLS_EXTENSION__,
        previousState: store.getState(),
        currentState: reduxStore.getState()
      });
    }
  });
}
