// react-imvc/hoc/connect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var returnNull = function returnNull() {
  return null;
};
var connect = function connect() {
  var selector = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : returnNull;
  return function (InputComponent) {
    function Connector(props, context) {
      return _react2.default.createElement(InputComponent, _extends({}, props, selector(_extends({}, context, { props: props }))));
    }

    Connector.contextTypes = {
      state: _react2.default.PropTypes.object,
      handlers: _react2.default.PropTypes.object,
      actions: _react2.default.PropTypes.object
    };

    return Connector;
  };
};

exports.default = connect;
